runtime/startup_byt.c

```diff
@@ -462,26 +459,11 @@ CAMLexport void caml_main(char_os **argv)
   char_os * shared_lib_path, * shared_libs;
   char_os * exe_name, * proc_self_exe;
 
-  /* Initialize the domain */
-  caml_init_domain();
-
-  /* Determine options */
-#ifdef DEBUG
-  caml_verb_gc = 0x3F;
-#endif
-  caml_parse_ocamlrunparam();
-  CAML_EVENTLOG_INIT();
-#ifdef DEBUG
-  caml_gc_message (-1, "### OCaml runtime: debug mode ###\n");
-#endif
-  if (!caml_startup_aux(/* pooling */ caml_cleanup_on_exit))
+  if (!caml_startup_aux(0)) {
```

Comment:
What is the point of this `pooling` parameter if it is always determined by the `caml_cleanup_on_exit` variable, either in the caller or the callee? I have the impression that we could either (1) get rid of it completely, or (2) never mutate it from within a function but always pass `caml_cleanup_on_exit` when entering the first function in a chain that is parametrized on `pooling`. Both would result in cleaner code than the odd mix we have today, and in particular I don't understand this `0` value here.

Reply:
My answer was going to be that it is actually the logical OR of caml_cleanup_on_exit (obtained from parsing OCAMLRUNPARAM) and the pooling parameter (non-zero in other callers of caml_startup_aux).

Looking more closely, your objection might be due to the fact that this does not make sense anyway because caml_cleanup_on_exit needs to be updated to correctly call caml_shutdown (bug that was there before): instead of 
```c
  if (caml_cleanup_on_exit)
    pooling = 1;
  if (pooling)
    caml_stat_create_pool();
```
it should be
```c
  if (pooled)
    caml_cleanup_on_exit = 1;
  if (caml_cleanup_on_exit)
    caml_stat_create_pool();
```
Currently, caml_startup_pooled does not release memory on shutdown (and it never did) when its was its sole purpose.

Anyway, grepping the opam repo shows that nobody uses caml_startup_pooled. (I also failed to find evidence that anyone uses the OCAMLRUNPARAM `c` but this is harder to make sure of.)

Reply:
> Anyway, grepping the opam repo shows that nobody uses caml_startup_pooled.

This may be a bit misleading because we expect these settings to be decided by end-user programs, not libraries, and it much more common to distribute support libraries on opam than programs. The propose use-case to introduce pooled startup (#71) was to support dynamic loading and unloading of OCaml plugins written for a "properietary Delphi software"; I would not expect to find those on opam!

Reply:
This is true. I know the caveats of grepping opam, and I had the usual precautions in my mind when I wrote that, but it is important to remind those caveats to not give false impressions. Thanks for the info as well.

Reply:
> Looking more closely, your objection might be due to the fact that this does not make sense anyway because caml_cleanup_on_exit **needs to** be updated to correctly call caml_shutdown

It doesn't need to.

Actually, `caml_cleanup_on_exit` (which is a state of a flag set via OCAMLRUNPARAM) is a superset of the pooling mode. Not only it enables pooling, but also enables shutdown on `exit` (which isn't desirable in all situations):

```c
CAMLprim value caml_sys_exit(value retcode_v)
{
  ...
  if (caml_cleanup_on_exit)
    caml_shutdown();
  ...
}
```


> (bug that was there before)
> [...]  
> Currently, caml_startup_pooled does not release memory on shutdown (and it never did) when its was its sole purpose.

This doesn't seem true. I haven't tested this code since 4.06 (which I use), but looking at it right now (before your patch), it seems quite obvious that the pool gets initialized in `caml_startup_pooled` (via `caml_startup_aux(1)`), and that `caml_shutdown` frees it thereafter (if you don't forget calling it). 

Please, double-check this in `trunk`. Perhaps you broke something with your patch.

Reply:
Indeed, your explanations make sense to me. I saw the lines above, did not have the right expected behaviour in mind, and misunderstood what they did. I will remove the corresponding change and if there is something to clarify in the documentation then I will write something about it (it could be entirely my own misreading).